src/main/ruby/core/posix.rb

```diff
@@ -63,7 +63,9 @@ def self.attach_function(method_name, native_name = method_name, argument_types,
           args[i] = str
         end
 
-        bound_func.call(*args)
+        result = bound_func.call(*args)
+        result = Truffle::Interop.from_java_string(Truffle::Interop.unbox(result)) if return_type == :string
```

Comment:
Using UNBOX here will assume the bytes have a UTF-8 encoding, which might be wrong for instance for `read()`/`write()`. All sequences of bytes are also not all valid UTF-8 strings, so then it's just impossible to encode in a j.l.String, unless both side are aware of it and use the char[] as a byte[] (but how can they communicate that if they also have to handle textual UTF-8 bytes?).
Additionally, this does 2 re-encodings of the String, so reading bytes directly is better.

We can automate the conversion if the `return_type` is String, it just has to use the logic below (or similar) and handle a returned NULL pointer.

Reply:
How's this do you think? (`.nil?` on NFI's objects does work.)

```ruby
        if return_type == :string
          if result.nil?
            nil
          else
            ptr = Rubinius::FFI::Pointer.new(Truffle::Interop.as_pointer(result))
            result = ptr.read_string_to_null
          end
        end
```

Reply:
Yes, that's good, that's what I had below in `getenv`.

Reply:
And yes it's quite nice that .nil? sends IS_NULL :)